### PR TITLE
Listener not called on RealmList

### PR DESCRIPTION
**Pull request by beeender**
- Add a test case to show issue described in #5507.
- Update OS to a31fab1f56a3a to solve part of the issue (issue with RealmList).
- Need https://github.com/realm/realm-object-store/issues/605 to solve
  the issue in RealmResults side.

### CODE REVIEW

File: realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java [hunk 1]
```diff
@@ -465,4 +479,78 @@ public void onChange(RealmResults<Dog> collection, @Nullable OrderedCollectionCh
...
+    // 1. Add listener to a collection
+    // A. change the collection in a background thread
+    // 2. Remove the listener
+    // 3. Add another listener
+    // 4. the listener added in step 3 should be triggered with change set in step A
+    @Test
+    @RunTestInLooperThread
+    public void addChangeListener_bug5507() throws InterruptedException {
+        // FIXME: See https://github.com/realm/realm-object-store/issues/605
+        if (type == ObservablesType.REALM_RESULTS) looperThread.testComplete();
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java]: I think you want to return here as well. As the rest of the code still runs I think.

File: realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java [hunk 2]
```diff
@@ -465,4 +479,78 @@ public void onChange(RealmResults<Dog> collection, @Nullable OrderedCollectionCh
...
+
+        OrderedRealmCollection<Dog> dogs = getTestingCollection(realm);
+        assertEquals(1, dogs.size());
+
+        if (type == ObservablesType.REALM_LIST) {
+            //noinspection unchecked
+            ((RealmList) dogs).addChangeListener(listener1);
+        } else {
+            //noinspection unchecked
+            ((RealmResults) dogs).addChangeListener(listener1);
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java]: Is this `else` statement ever reached?

File: realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java [hunk 3]
```diff
@@ -465,4 +479,78 @@ public void onChange(RealmResults<Dog> collection, @Nullable OrderedCollectionCh
...
+        }
+
+
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java]: Whitespace

File: realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java (hunk 2 above)
Comment (cmelchior) [reply realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java]: Or are you just keeping them here so the test case is ready for the FIXME?

File: realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java [hunk 4]
```diff
@@ -465,4 +479,78 @@ public void onChange(RealmResults<Dog> collection, @Nullable OrderedCollectionCh
...
+    // To reproduce https://github.com/realm/realm-java/issues/5507
+    // 1. Add listener to a collection
+    // A. change the collection in a background thread
+    // 2. Remove the listener
+    // 3. Add another listener
+    // 4. the listener added in step 3 should be triggered with change set in step A
+    @Test
+    @RunTestInLooperThread
+    public void addChangeListener_bug5507() throws InterruptedException {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java]: Awesome job finding a unit test that reproduces this 💯

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,7 @@
 
 * Added missing `toString()` for the implementation of `OrderedCollectionChangeSet`.
 * Sync queries are evaluated immediately to solve the performance issue when the query results are huge, `RealmResults.size()` takes too long time (#5387).
+* Fixed the listener on `RealmList` was not called issue when removing the listener then adding it again (#5507). Please notice the similar issue still exists for `RealmResults`.
```
Comment (cmelchior) [inline CHANGELOG.md]: was not called issue when removing -> not being called when removing
Please notice that a similar issue

File: realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java (hunk 2 above)
Comment (beeender) [reply realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java]: yeah, make it ready for the FIXME. Added return statement at the beginning.

**Pull request merged**